File: src/leetcodepython/array/find_continuous_sequence_57.py
# -*- coding:utf-8 -*-
'''
/**
 * This is the solution of No. 57 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/insert-interval
 *
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给出一个无重叠的 ，按照区间起始端点排序的区间列表。
 *
 * 在列表中插入一个新的区间，你需要确保列表中的区间仍然有序且不重叠（如果有必要的话，可以合并区间）。
 *
 * 示例 1:
 *
 * 输入: intervals = [[1,3],[6,9]], newInterval = [2,5]
 * 输出: [[1,5],[6,9]]
 * 示例 2:
 *
 * 输入: intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]], newInterval = [4,8]
 * 输出: [[1,2],[3,10],[12,16]]
 * 解释: 这是因为新的区间 [4,8] 与 [3,5],[6,7],[8,10] 重叠。
 *
 * 来源：力扣（LeetCode）
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
'''
from typing import List


def find_continuous_sequence(target: int) -> List[List[int]]:
    '''
        查找连续的序列
    Args:
        target: 目标值
    Returns:
        目标数组
    '''
    end = int(target / 2) + 1
    result = []
    for i in range(1, end):
        for j in range(i + 1, end + 1):
            total = (i + j) * (j - i + 1) / 2
            if total == target:
                result.append([x for x in range(i, j + 1)])
                break
    return result


def find_continuous_sequence2(target: int) -> List[List[int]]:
    '''
        查找连续的序列
    Args:
        target: 目标值
    Returns:
        目标数组
    '''
    result = []
    left,right = 1,2
    while left < right:
        total = (left + right) * (right - left + 1) / 2
        if total < target:
            right += 1
        elif total > target:
            left += 1
        else:
            result.append([x for x in range(left, right + 1)])
            left += 1
            right += 1
    return result


if __name__ == '__main__':
    result = find_continuous_sequence2(50252)
    print(result)

File: src/leetcodejava/array/FindContinuousSequence57.java
package leetcodejava.array;

import org.junit.Assert;
import org.junit.Test;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.IntStream;

/**
 * This is the solution of No. offer57 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/he-wei-sde-lian-xu-zheng-shu-xu-lie-lcof
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 输入一个正整数 target ，输出所有和为 target 的连续正整数序列（至少含有两个数）。
 * <p>
 * 序列内的数字由小到大排列，不同序列按照首个数字从小到大排列。
 * <p>
 * 示例 1：
 * <p>
 * 输入：target = 9
 * 输出：[[2,3,4],[4,5]]
 * 示例 2：
 * <p>
 * 输入：target = 15
 * 输出：[[1,2,3,4,5],[4,5,6],[7,8]]
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/he-wei-sde-lian-xu-zheng-shu-xu-lie-lcof
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class FindContinuousSequence57 {

    @Test
    public void findContinuousSequenceTest() {
        int target = 9;
        int[][] result = findContinuousSequence3(target);
        System.out.println(result);
        Assert.assertEquals(result.length, 3);
    }

    /**
     * 连续数组和为固定值
     *
     * @param target 固定值
     * @return 连续数组
     */
    public int[][] findContinuousSequence1(int target) {
        List<int[]> result = new ArrayList<>();
        int end = target / 2 + 1;
        for (int i = 1; i < end; i++) {
            for (int j = i + 1; j < end + 1; j++) {
                int total = (i + j) * (j - i + 1) / 2;
                if (target == total) {
                    result.add(IntStream.range(i, j + 1).toArray());
                    break;
                }
            }
        }
        return result.stream().toArray(int[][]::new);
    }

    /**
     * 连续数组和为固定值
     *
     * @param target 固定值
     * @return 连续数组
     */
    public int[][] findContinuousSequence2(int target) {
        int left = 1;
        int right = 2;
        List<int[]> result = new ArrayList<>();
        while (left < right) {
            int total = (left + right) * (right - left + 1) / 2;
            if (total > target) {
                left++;
            } else if (total < target) {
                right++;
            } else {
                result.add(IntStream.range(left, right + 1).toArray());
                left++;
                right++;
            }
        }
        return result.stream().toArray(int[][]::new);
    }

    /**
     * 连续数组和为固定值
     *
     * @param target 固定值
     * @return 连续数组
     */
    public int[][] findContinuousSequence3(int target) {
        List<List<Integer>> result = new ArrayList<>();
        int sum = 0;
        List<Integer> temp = new LinkedList<>();
        int i = 1;
        while (i < target) {
            if (sum < target) {
                temp.add(i);
                sum += i;
                i++;
            } else if (sum == target) {
                result.add(new ArrayList<>(temp));
                temp.add(i);
                sum += i;
                i++;
            } else {
                sum -= temp.remove(0);
            }
        }
        int size = result.size();
        int res[][] = new int[size][];
        int index = 0;
        for (List<Integer> e : result) {
            int len = e.size();
            int[] pair = new int[len];
            for (int k = 0; k < len; k++) {
                pair[k] = e.get(k);
            }
            res[index++] = pair;
        }
        return res;
    }
}
